wolfcrypt/src/port/psa/psa_pkcbs.c

```diff
@@ -0,0 +1,554 @@
+
+    psa_set_key_type(&attr, PSA_KEY_TYPE_ECC_PUBLIC_KEY(ecc_curve));
+    psa_set_key_usage_flags(&attr, PSA_KEY_USAGE_VERIFY_HASH);
+    psa_set_key_bits(&attr, ecc_curve_size * 8);
+    psa_set_key_algorithm(&attr, PSA_ALG_ECDSA_ANY);
+
+    PSA_LOCK();
+    status = psa_import_key(&attr, raw_key, raw_key_length, key_id);
+    PSA_UNLOCK();
+    if (status != PSA_SUCCESS) {
+        WOLFSSL_MSG("psa_ecc_decode_public_key: can't import the key");
+        ret = WC_HW_E;
+        goto exit;
+    }
+
+    ret = 0;
+
+exit:
+    wc_ecc_free(&tmp_key);
+    return ret;
+}
+
+static int psa_ecc_verify_cb(WOLFSSL* ssl, const byte* sig, word32 sig_length,
+                             const byte* hash, word32 hash_length,
+                             const byte* key, word32 key_length,
+                             int* result, void* ctx)
+{
+    uint8_t raw_signature[ECC_MAXSIZE * 2], *s;
+    psa_key_id_t tmp_key;
+    word32 r_len, s_len;
+    psa_status_t status;
+    int ret;
+
+    (void)ssl;
+    (void)ctx;
+    WOLFSSL_ENTER("psa_ecc_verify_cb");
+
+    ret = psa_ecc_decode_public_key(key, key_length, &tmp_key);
+    if (ret != 0)
+        return ret;
+
+    r_len = s_len = ECC_MAXSIZE;
+    s = raw_signature + r_len;
+    ret = wc_ecc_sig_to_rs(sig, sig_length, raw_signature, &r_len,
+                           s, &s_len);
+    if (ret != 0)
+        goto exit;
+
+    /* coalescence of r and s in the buffer */
+    XMEMCPY(raw_signature + r_len, s, s_len);
+
+    PSA_LOCK();
+    status = psa_verify_hash(tmp_key, PSA_ALG_ECDSA_ANY, hash,
+                             hash_length, raw_signature, r_len + s_len);
+    PSA_UNLOCK();
+
+    if (status == PSA_SUCCESS) {
+        *result = 1;
+    } else {
+        *result = 0;
+
+        if (status != PSA_ERROR_INVALID_SIGNATURE) {
+            WOLFSSL_MSG("psa_ecc_verify_cb: can't verify hash");
+            ret = WC_HW_E;
+        }
+
+        /* returns 0 for invalid signature as well */
+    }
+
+ exit:
+    PSA_LOCK();
+    psa_destroy_key(tmp_key);
+    PSA_UNLOCK();
+
+    return ret;
+}
+#endif
+
+#ifdef HAVE_HKDF
+static int psa_hkdf_extract_cb(byte* prk, const byte* salt,
+                               word32 salt_length, byte* ikm,
+                               word32 ikm_length, int digest,
+                               void* ctx)
+{
+    psa_key_attributes_t attr = { 0 };
+    psa_algorithm_t hash;
+    psa_key_id_t tmp_key;
+    psa_status_t status;
+    size_t digest_len;
+
+    (void)ctx;
+
+    WOLFSSL_ENTER("psa hkdf cb");
+
+    if (salt_length != 0 && salt == NULL)
+        return BAD_FUNC_ARG;
+
+    switch (digest) {
+#ifndef NO_SHA256
+        case WC_SHA256:
+            digest_len = WC_SHA256_DIGEST_SIZE;
+            hash = PSA_ALG_SHA_256;
+            break;
+#endif
+
+#ifdef WOLFSSL_SHA384
+        case WC_SHA384:
+            digest_len = WC_SHA384_DIGEST_SIZE;
+            hash = PSA_ALG_SHA_384;
+            break;
+#endif
+
+#ifdef WOLFSSL_TLS13_SHA512
```

Comment:
Interesting... `WOLFSSL_TLS13_SHA512`, that's new, but I see how it is used. For PK callbacks this looks fine. I would have used `WOLFSSL_SHA512`

Reply:
You're right. This should be WOLFSSL_SHA512

Reply:
AFAIU SHA512 is not supported in any current cipher suite for HKDF, so I'm removing it.